src/server/auth/server/api_server.go

```diff
@@ -2802,3 +2802,120 @@ func (a *apiServer) SetConfiguration(ctx context.Context, req *auth.SetConfigura
 	}
 	return &auth.SetConfigurationResponse{}, nil
 }
+
+func (a *apiServer) ExtractAuthTokens(ctx context.Context, req *auth.ExtractAuthTokensRequest) (resp *auth.ExtractAuthTokensResponse, retErr error) {
+	switch a.activationState() {
+	case none:
+		return nil, auth.ErrNotActivated
+	case partial:
+		return nil, auth.ErrPartiallyActivated
+	}
+
+	// We don't want to actually log the request/response since they contain
+	// credentials.
+	defer func(start time.Time) { a.LogResp(nil, nil, retErr, time.Since(start)) }(time.Now())
+
+	// Get calling user. The user must be an admin to extract auth tokens.
+	callerInfo, err := a.getAuthenticatedUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	isAdmin, err := a.hasClusterRole(ctx, callerInfo.Subject, auth.ClusterRole_SUPER)
+	if err != nil {
+		return nil, err
+	}
+	if !isAdmin {
+		return nil, &auth.ErrNotAuthorized{
+			Subject: callerInfo.Subject,
+			AdminOp: "ExtractAuthTokens",
+		}
+	}
+
+	extracted := make([]*auth.HashedAuthToken, 0)
+
+	tokens := a.tokens.ReadOnly(ctx)
+	var val auth.TokenInfo
+	if err := tokens.List(&val, col.DefaultOptions, func(hash string) error {
+		// Only extract robot tokens
+		if !strings.HasPrefix(val.Subject, auth.RobotPrefix) {
+			return nil
+		}
+
+		ttl, err := tokens.TTL(hash)
+		if err != nil {
+			return err
+		}
+
+		token := &auth.HashedAuthToken{
+			HashedToken: hash,
+			TokenInfo: &auth.TokenInfo{
+				Subject: val.Subject,
+				Source:  val.Source,
+			},
+		}
+		if ttl != -1 {
+			expiration, err := types.TimestampProto(time.Now().Add(time.Second * time.Duration(ttl)))
```

Comment:
This feels weird, maybe we should put the absolute expiration timestamp in the TokenInfo in 2.0?

Reply:
I originall wrote this for the comment above, but moved it here since this seems like the more relevant place.

In you view, is the grossness due to skew (i.e. the TTL you get may not be the TTL anymore after getting the current time and adding the two, etc)? There's a bit of a limitation that comes from etcd's lease API dealing in durations, but we could always work around that by checking the token's expiration and disregarding it if it's in the past, even if the etcd lease hasn't expired yet.

Reply:
Yup, the introduction of skew feels weird. I don't think we need to address it here,  but like you said in the future we could do a belt-and-suspenders by setting a TTL and also checking a timestamp encoded into the token info.

AFAIK this PR should address any issues with trying to restore expired tokens.